---
title: "Reference library coverage report"
author: "Rupert A. Collins"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: github_document
---


##### Methods and description
This document describes the contents of the UK fish reference library, generated from public databases. 
The document is a dynamic knitr document and can be updated quickly using the Makefile in `scripts/`.
A list of species from the UK was generated from three sources: GBIF, FishBase, and the Water Framework Directive list of transitional species.
This list was filtered to identify synonyms and duplicates, and annotated with FishBase taxonomic classification and FishBase common names.
Next a sub-list of "common" species was generated. 
These were species that we believe are likely to be encountered in eDNA surveys of inshore and transitional waters of the UK, and comprise most of the species in Henderson (2015).
Most of the remaining are either introduced species, rarely encountered migrants, oceanic pelagics, or deep sea organisms.


The search was performed on the GenBank nt and BOLD sequences databases. 
Because of inconsistencies in how researchers annotate their GenBank submissions and the differing internal coverage of primer pairs for particular gene fragments, we performed a search requesting all mitochondrial DNA.
Then we pulled out fragments of interest using a hidden markov model. This enabled us to have greater confidence that useful sequences had not been missed.
For the resulting sequences we then tabulated all their metadata from GenBank in order to allow us the capability to curate a custom reference library according to various criteria (e.g. must have reference specimen or locality data).


```{r load_libs, include=FALSE}
# load up libs
source("load-libs.R")
# source("scripts/load-libs.R")
```


```{r load_data, include=FALSE}
# load up the data
source("references-load.R")
# source("scripts/references-load.R")
genbank.df <- reflib.orig
activity.df <- read_csv(file="../reports/stats.csv")
# activity.df <- read_csv(file="reports/stats.csv")
```


```{r activities, include=FALSE}
# get dates and genbank version
gb.version <- activity.df %>% filter(stat=="genbankVersion") %>% pull(n)
gb.date <- activity.df %>% filter(stat=="date") %>% pull(n)
```


```{r species_clean, include=FALSE}
# clean up the species data
uk.species.table.val <- uk.species.table %>% rename(sciNameValid=validName)
```


```{r merge_df, include=FALSE}
# count sequences per species-marker
reflib.by.marker <- reflib.orig %>% 
    select(sciNameValid,starts_with("nucleotidesFrag")) %>%
    full_join(uk.species.table.val,by="sciNameValid") %>%
    pivot_longer(cols=!c(class,order,family,genus,commonName,commonSpecies,sciNameValid),names_to="marker",values_to="nucleotides") %>% 
    mutate(hasSeq=if_else(is.na(nucleotides),0,1)) %>% 
    group_by(class,order,family,genus,commonName,commonSpecies,sciNameValid,marker) %>%
    summarise(genbankCount=sum(hasSeq),.groups="drop") %>%
    mutate(marker=str_replace_all(marker,"nucleotidesFrag.",""), marker=str_replace_all(marker,".noprimers",""))
```


```{r get_haps, include=FALSE}
# get the prefixes 
prefixes <- reflib.orig %>% select(starts_with("nucleotidesFrag")) %>% names() %>% str_replace_all("nucleotidesFrag\\.","")
# subset each marker
reflibs.sub <- mcmapply(function(x) subset_nucs(pref=x,df=reflib.orig), prefixes, SIMPLIFY=FALSE,USE.NAMES=TRUE,mc.cores=2)
# collapse dataframe by haps-per-species, annotate with number haps
reflibs.haps <- mcmapply(function(x) haps2fas(df=x), reflibs.sub, SIMPLIFY=FALSE,USE.NAMES=TRUE,mc.cores=2)
# summarise
haps.table <- mapply(FUN=function(x,y) x %>% select(sciNameValid) %>% count(sciNameValid) %>% mutate(marker=y), reflibs.haps, names(reflibs.haps), SIMPLIFY=FALSE) %>% purrr::reduce(rbind) %>% mutate(marker=str_replace_all(marker,".noprimers",""))
```


```{r summary_tab, include=FALSE}
# reformat
marker.table <- reflib.by.marker %>% 
    group_by(marker) %>% 
    summarise(
            `Total`=sum(genbankCount),
            `Cov. (all)`=length(which(genbankCount>0))/length(genbankCount),
            `Cov. (common)`=length(which(genbankCount[which(commonSpecies==TRUE)]>0))/length(genbankCount[which(commonSpecies==TRUE)]),
            `Cov. (rare)`=length(which(genbankCount[which(commonSpecies==FALSE)]>0))/length(genbankCount[which(commonSpecies==FALSE)]),
            `Singletons`=length(which(genbankCount==1))/length(which(genbankCount>0))
            ) 
# format haplotype table
haps.format <- haps.table %>% group_by(marker) %>% summarise(`Haps (mean)`=mean(n),`Haps (median)`=median(n),.groups="drop")
# join and format
marker.table.pretty <- marker.table %>% left_join(haps.format) %>% separate(marker,into=c("Locus","Fragment"))
```


```{r percentages, include=FALSE}
# calculate percentages for each group
uk.all <- reflib.by.marker %>% count()
```


##### Results

The total number of UK species is estimated to be around `r uk.all`.
GenBank and BOLD were searched on `r gb.date` (GenBank version `r gb.version`).


###### Table 1. Summary
```{r print_summary, echo=FALSE, results="asis"}
options(knitr.kable.NA="",digits=2)
# print the common species table
marker.table.pretty %>%
    mutate(Locus=str_to_upper(Locus),Fragment=str_to_title(Fragment)) %>%
    kable()
```


###### Table 2. Big table
```{r print_big, echo=FALSE, results="asis"}
options(knitr.kable.NA="")
# print the common species table
reflib.by.marker %>%
    mutate(locus=str_to_upper(str_split_fixed(marker,"\\.",2)[,1])) %>% 
    mutate(fragment=str_to_title(str_split_fixed(marker,"\\.",2)[,2])) %>%
    mutate(marker=paste0(locus," (",fragment,")")) %>%
    select(-locus,-fragment) %>%
    pivot_wider(names_from=marker,values_from=genbankCount,values_fill=0) %>%
    arrange(class,order,family,sciNameValid) %>%
    filter(commonSpecies==TRUE) %>%
    select(-class,-order,-commonSpecies) %>%
    mutate(sciNameValid=str_replace_all(sciNameValid, pattern="$|^", replacement="*")) %>%
    rename(Family=family, `Scientific Name`=sciNameValid, `Common Name`=commonName) %>%
    kable()
```
